File: android/app/src/main/java/net/kaltner/palomar/RepositoryLabels.kt
package net.kaltner.palomar

internal data class ResolvedRepository(
    val info: RepositoryInfo,
    val canonicalPath: String,
    val label: String,
)

private data class RepositoryLabelCandidate(
    val canonicalPath: String,
    val name: String,
    val parentSegments: List<String>,
    var depth: Int = 1,
)

internal fun resolveRepositorySet(
    repositories: List<RepositoryInfo>,
    repositoryRoot: String,
): List<ResolvedRepository> {
    val candidates = repositories.map { repository ->
        val relativePath = repositoryRelativePath(repository.path, repositoryRoot)
        val segments = if (relativePath == ".") emptyList() else relativePath.split('/').filter(String::isNotEmpty)
        RepositoryLabelCandidate(
            canonicalPath = canonicalRepositoryPath(repository.path, repositoryRoot),
            name = repository.name.ifBlank { segments.lastOrNull() ?: repository.path },
            parentSegments = segments.dropLast(1),
        )
    }

    while (true) {
        var expanded = false
        candidates.indices.groupBy { candidateLabel(candidates[it]) }.values.forEach { indexes ->
            if (indexes.mapTo(linkedSetOf()) { candidates[it].canonicalPath }.size < 2) return@forEach
            indexes.forEach { index ->
                val candidate = candidates[index]
                if (candidate.depth <= candidate.parentSegments.size) {
                    candidate.depth += 1
                    expanded = true
                }
            }
        }
        if (!expanded) break
    }

    return repositories.mapIndexed { index, repository ->
        ResolvedRepository(repository, candidates[index].canonicalPath, candidateLabel(candidates[index]))
    }
}

internal fun matchResolvedRepository(
    path: String,
    repositories: List<ResolvedRepository>,
): ResolvedRepository? {
    val cwd = normalizeRepositoryPath(path)
    return repositories
        .filter { cwd == it.canonicalPath || cwd.startsWith("${it.canonicalPath}/") }
        .maxByOrNull { it.canonicalPath.length }
}

internal fun normalizeRepositoryPath(value: String): String =
    value.trimEnd('/').ifBlank { "/" }

private fun candidateLabel(candidate: RepositoryLabelCandidate): String {
    if (candidate.depth == 1) return candidate.name
    val parentCount = candidate.depth - 1
    return (candidate.parentSegments.takeLast(parentCount) + candidate.name).joinToString("/")
}

private fun canonicalRepositoryPath(path: String, repositoryRoot: String): String =
    normalizeRepositoryPath(
        if (path.startsWith('/')) path
        else "${repositoryRoot.trimEnd('/')}/$path",
    )

private fun repositoryRelativePath(path: String, repositoryRoot: String): String {
    val normalizedPath = normalizeRepositoryPath(path)
    if (!path.startsWith('/')) return normalizedPath.removePrefix("./")
    val normalizedRoot = normalizeRepositoryPath(repositoryRoot)
    return when {
        normalizedPath == normalizedRoot -> "."
        normalizedPath.startsWith("$normalizedRoot/") -> normalizedPath.removePrefix("$normalizedRoot/")
        else -> normalizedPath.trimStart('/')
    }
}

File: web/src/repository-labels.ts
import type { RepositoryInfo } from "./protocol";

export interface ResolvedRepository {
  info: RepositoryInfo;
  canonicalPath: string;
  label: string;
}

interface LabelCandidate {
  canonicalPath: string;
  name: string;
  parentSegments: string[];
  depth: number;
}

export function resolveRepositorySet(
  repositories: RepositoryInfo[],
  repositoryRoot: string,
): ResolvedRepository[] {
  const candidates = repositories.map((repository): LabelCandidate => {
    const relativePath = repositoryRelativePath(repository.path, repositoryRoot);
    const segments = relativePath === "." ? [] : relativePath.split("/").filter(Boolean);
    return {
      canonicalPath: canonicalRepositoryPath(repository.path, repositoryRoot),
      name: repository.name || segments.at(-1) || repository.path,
      parentSegments: segments.slice(0, -1),
      depth: 1,
    };
  });

  while (true) {
    const collisions = new Map<string, number[]>();
    candidates.forEach((candidate, index) => {
      const label = candidateLabel(candidate);
      collisions.set(label, [...(collisions.get(label) ?? []), index]);
    });
    let expanded = false;
    collisions.forEach((indexes) => {
      const canonicalPaths = new Set(indexes.map((index) => candidates[index].canonicalPath));
      if (canonicalPaths.size < 2) return;
      indexes.forEach((index) => {
        const candidate = candidates[index];
        if (candidate.depth <= candidate.parentSegments.length) {
          candidate.depth += 1;
          expanded = true;
        }
      });
    });
    if (!expanded) break;
  }

  return repositories.map((info, index) => ({
    info,
    canonicalPath: candidates[index].canonicalPath,
    label: candidateLabel(candidates[index]),
  }));
}

export function matchResolvedRepository(
  path: string,
  repositories: ResolvedRepository[],
): ResolvedRepository | undefined {
  const cwd = normalizeRepositoryPath(path);
  return repositories
    .filter(({ canonicalPath }) => cwd === canonicalPath || cwd.startsWith(`${canonicalPath}/`))
    .sort((left, right) => right.canonicalPath.length - left.canonicalPath.length)[0];
}

export function repositoryDisplayName(
  path: string,
  repositories: ResolvedRepository[],
): string {
  return matchResolvedRepository(path, repositories)?.label ?? shortPathName(path);
}

export function normalizeRepositoryPath(value: string): string {
  return value.replace(/\/+$/, "") || "/";
}

function candidateLabel(candidate: LabelCandidate): string {
  if (candidate.depth === 1) return candidate.name;
  const parentCount = candidate.depth - 1;
  return [...candidate.parentSegments.slice(-parentCount), candidate.name].join("/");
}

function canonicalRepositoryPath(path: string, repositoryRoot: string): string {
  return normalizeRepositoryPath(path.startsWith("/")
    ? path
    : `${repositoryRoot}/${path}`);
}

function repositoryRelativePath(path: string, repositoryRoot: string): string {
  const normalizedPath = normalizeRepositoryPath(path);
  if (!path.startsWith("/")) return normalizedPath.replace(/^\.\//, "");
  const normalizedRoot = normalizeRepositoryPath(repositoryRoot);
  if (normalizedPath === normalizedRoot) return ".";
  if (normalizedPath.startsWith(`${normalizedRoot}/`)) {
    return normalizedPath.slice(normalizedRoot.length + 1);
  }
  return normalizedPath.replace(/^\/+/, "");
}

function shortPathName(path: string): string {
  if (!path || path === "(no repository)") return "No repository";
  return path.replace(/\/$/, "").split("/").at(-1) || path;
}
